controller/src/results.rs

```diff
@@ -247,6 +248,62 @@ impl<P> ModuleResult<P> {
     }
 }
 
+impl<P: Clone> ModuleResult<P> {
+    /// Merge the successful and error data from one result into `self`.
+    ///
+    /// This will take all the successful module IDs and data, and merge them
+    /// into `self`'s success. Any failures in `other` will be merged into
+    /// `self.failures`.
+    ///
+    /// This returns `None` if any of the following occur:
+    ///
+    /// - `other.modules` and `self.modules` overlap. In this case, it is not
+    /// possible to build a consistent result, since we have two values for the
+    /// successful data.
+    /// - `other.failures` and `self.failures` overlap, for the same reason.
+    ///
+    /// Note that if a module exists in `other.failures.modules` and
+    /// `self.modules`, then it is _removed_ from `self.modules` and added to
+    /// the failures. The converse is _not_ true: if a module exists in
+    /// `self.failures.modules` and `other.modules` it is _not_ removed. That
+    /// is, failures are "sticky" in `self`.
+    pub fn merge(&self, other: &Self) -> Option<Self> {
+        // The intersection of the successful module IDs should be empty.
+        if (self.modules.0 & other.modules.0) != 0 {
+            return None;
+        }
+
+        // Same for the failed module IDs.
+        if (self.failures.modules.0 & other.failures.modules.0) != 0 {
+            return None;
+        }
+
+        // Remove any successes we currently have that are in the new failures.
+        let (remaining_successes, remaining_data) =
+            remove_module_data(self.modules, self.data.iter(), other.failures.modules);
```

Comment:
Nit: this could be a function on `self`, since we pass `self.modules` and `self.data` into it:
```rust
let (remaining_successes, remaining_data) =
            self.remove_module_data(other.failures.modules);
```
(I'm not sure of the naming, because `remove_module_data` suggests a function that modifies `&mut self`; maybe `without_module_data` and annotate with `#[must_use]`?)

Reply:
It could be, I think, but it's defined in `transceiver-messages`, not `transceiver-controller`, which is where `Self` (`ModuleResult`) is defined. I actually can't remember why I did that originally, though I guess it was that I thought I'd need them in that crate. I don't believe we use them there now. Is it worth schlepping them around now?

Reply:
It looks like we do use the free function elsewhere, so I'm suggesting adding a new function in `ModuleResult`:
```rust
    /// Removes failures in `other.failures` from `self.modules` and `self.data`
    fn without_failures_from(&self, other: Self) -> (ModuleId, Vec<P>) {
        remove_module_data(self.modules, self.data.iter(), other.failures.modules)
    }
```

I don't feel too strongly about it, though.

Reply:
Ah, got it. Yeah that'd be convenient, I've been a bit worried by the fact that we can call those methods with a bitmask and data that may not be consistent. Adding now.

Reply:
Done in a38b0c7